src/System.Collections.Immutable/tests/ImmutableArrayTest.cs

```diff
@@ -479,630 +577,1045 @@ public void LastIndexOf()
+                Assert.Throws<InvalidOperationException>(() => ((IImmutableList<int>)s_emptyDefault).RemoveRange(s_emptyDefault, comparer));
+
+                // Enumerable overloads, rhs is null
+                Assert.Throws<ArgumentNullException>("items", () => array.RemoveRange(items: null));
+                Assert.Throws<ArgumentNullException>("items", () => array.RemoveRange(items: null, equalityComparer: comparer));
+                Assert.Throws<ArgumentNullException>("items", () => ((IImmutableList<int>)array).RemoveRange(items: null));
+                Assert.Throws<ArgumentNullException>("items", () => ((IImmutableList<int>)array).RemoveRange(items: null, equalityComparer: comparer));
+
+                // Enumerable overloads, lhs is default and rhs is null
+                Assert.Throws<NullReferenceException>(() => s_emptyDefault.RemoveRange(items: null));
+                Assert.Throws<NullReferenceException>(() => s_emptyDefault.RemoveRange(items: null, equalityComparer: comparer));
+                Assert.Throws<InvalidOperationException>(() => ((IImmutableList<int>)s_emptyDefault).RemoveRange(items: null));
+                Assert.Throws<InvalidOperationException>(() => ((IImmutableList<int>)s_emptyDefault).RemoveRange(items: null, equalityComparer: comparer));
+            });
+        }
+
+        [Fact]
+        public void RemoveRangeEnumerableRegression()
+        {
+            // Validates that a fixed bug in the inappropriate adding of the Empty
+            // singleton enumerator to the reusable instances bag does not regress.
+            
+            IEnumerable<int> oneElementBoxed = s_oneElement;
+            IEnumerable<int> emptyBoxed = s_empty;
+            IEnumerable<int> emptyDefaultBoxed = s_emptyDefault;
+
+            Assert.Throws<NullReferenceException>(() => s_emptyDefault.RemoveRange(emptyBoxed));
+            Assert.Throws<NullReferenceException>(() => s_emptyDefault.RemoveRange(emptyDefaultBoxed));
+            Assert.Throws<InvalidOperationException>(() => s_empty.RemoveRange(emptyDefaultBoxed));
+
+            Assert.Equal(oneElementBoxed, oneElementBoxed);
         }
 
-        [Fact]
-        public void SetItem()
+        [Theory]
+        [MemberData(nameof(RemoveAllData))]
+        public void RemoveAll(IEnumerable<int> source, Predicate<int> match)
         {
-            Assert.Throws<ArgumentOutOfRangeException>("index", () => s_empty.SetItem(0, 10));
-            Assert.Throws<NullReferenceException>(() => s_emptyDefault.SetItem(0, 10));
-            Assert.Throws<ArgumentOutOfRangeException>("index", () => s_oneElement.SetItem(1, 10));
-            Assert.Throws<ArgumentOutOfRangeException>("index", () => s_empty.SetItem(-1, 10));
+            var array = source.ToImmutableArray();
+            var expected = source.Where(i => !match(i));
+            Assert.Equal(expected, array.RemoveAll(match));
+        }
 
-            Assert.Equal(new[] { 12345 }, s_oneElemen
```

Comment:
This was the one where the `IImmutableList` box was causing an IOE to be thrown for default immutable arrays.